### ISSUE TITLE
Add a method to save and load the optimizer.

### ISSUE THREAD

**Issue by w4nderlust** [tensorflow/tensorflow#41053]
In TensorFlow 2.2 there is the capability to save a model with its optimizer.
This is fine, but I run in situations where I want to save the weights and the optimizer separately (conflating training, metrics architecture and weights in a single model object is a bit confusing too imho).
If I try to pickle the optimizer and then reload it, it looks like the contents of the the loaded optimizer are identical to the previous one, but when training the losses are different.

I tested this by creating a model, training it for 4 epochs, collecting losses and its predictions, then training a model for 2 epochs, saving weights and pickling the optimizer, reloading it and training for 2 more epochs. I  reset the random seed where appropriate.
What happens is that, if the optimizer is stateless like SGD, the predictions and all losses are identical (so you can pickle and unpickle a stateless optimizer, good!), but if I try to use a stateful optimizer like Adam,  losses and predictions differ (although all the variables are loaded correclty, as seen from manual inspection).
This suggests that there is something missing somewhere in my code that is not obvious. A save and load mechanism for the optimizer that takes care of whatever else is needed would be greatly appreciated.

Here's my code for you to replicate the issue.

```python
import pickle

import numpy as np
import tensorflow as tf
from tensorflow.keras import Model
from tensorflow.keras.layers import Dense
from tensorflow.keras.losses import MeanSquaredError
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.optimizers import Adam

#optimizer_type = SGD  # this works
optimizer_type = Adam  # this does not work

np.random.seed(42)

x = np.array(range(100), dtype=np.float32).reshape(-1, 1)
x_batched = np.split(x, 2)
y = 2 * x + 1 + np.random.normal(size=x.shape[0]).reshape(-1, 1)
y_batched = np.split(y, 2)

loss_fn = MeanSquaredError()


class MyModel(Model):
    def __init__(self):
        super().__init__()
        self.d1 = Dense(5, activation='relu', dtype=tf.float32)
        self.d2 = Dense(1, dtype=tf.float32)

    def call(self, inputs, **kwargs):
        return self.d2(self.d1(inputs))


def train_step(model, loss_fn, optimizer, x, y):
    with tf.GradientTape() as tape:
        logits = model(x, training=True)
        loss_value = loss_fn(y, logits)
    grads = tape.gradient(loss_value, model.trainable_weights)
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
    return loss_value


def optimize(model, loss_fn, optimizer, epochs):
    losses = []
    for epoch in range(epochs):
        for x_batch, y_batch in zip(x_batched, y_batched):
            loss_value = train_step(model, loss_fn, optimizer, x_batch,
                                    y_batch)
            losses.append(loss_value.numpy())
    return losses


# train 4 epochs

tf.random.set_seed(42)
model = MyModel()
optimizer = optimizer_type(learning_rate=1e-3)
losses = optimize(model, loss_fn, optimizer, epochs=4)
pred = model(x_batched[0])

# train 2 epochs

tf.random.set_seed(42)
model = MyModel()
optimizer = optimizer_type(learning_rate=1e-3)
losses2 = optimize(model, loss_fn, optimizer, epochs=2)
model.save_weights('weights')
with open('opt.pkl', 'wb') as f:
    pickle.dump(optimizer, f)

# load and train 2 more epochs

model = MyModel()
model.load_weights('weights')
with open('opt.pkl', 'rb') as f:
    optimizer = pickle.load(f)
losses2.extend(optimize(model, loss_fn, optimizer, epochs=2))
pred2 = model(x_batched[0])

print(losses)
print(losses2)
print(pred)
print(pred2)

assert losses == losses2
assert np.all(pred == pred2)

```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/b82a2f337af07c3adbee0dd7ce547ab8/41053.ipynb). Thanks!

**Comment by marcus302** [comment]
Hello! Does somebody know what the status of this issue is? I'm running into exactly this problem.

**Comment by pandrey-fr** [comment]
Hi,

The issue here is that you only pickle the optimizer's instantiation parameters (which, by the way, are better accessed through the `get_config` / `from_config` API), not its stateful variables. Those can be accessed _via_ the `get_weights` / `set_weights` API.

That being said, there is still the problem of re-creating the weights in a restored optimizer before being able to assign them values. _I.e._ it is possible to do `wgt = model.optimizer.get_weights()` after some iterations and to reset the optimizer's states using `model.optimizer.set_wights(wgt)` after some iterations, but no to assign those weights (save for the number of past iterations) right at instantiation time, which is indeed quite an issue.

**Comment by w4nderlust** [comment]
@pandrey-fr thanks for your insight, I understand what you are saying. I believe a function pair that saves both init and stateful parameters and is able to reload them would be greatly appreciated.

**Comment by thatdc** [comment]
Any news on this feature? I cannot save the full model for different reasons and I must save the weights + the optimizer state (in my case SGD with decay) to continue the training process starting from a checkpoint.
I cannot figure out how to overcome the problem of saving the optimizer state without saving the whole model

**Comment by mattroos** [comment]
I'd very much like to have this capability as well.

**Comment by ooshyun** [comment]
@w4nderlust I also had same problem when not using checkpoint, but if I use tf.train.Checkpoint and tf.train.CheckpointManager, then the optimizer saved. I attached my test code, please check in this way ! 

Here's my code.

```python
import pickle

import numpy as np
import tensorflow as tf
from keras import Model
from keras.layers import Dense, Layer
from keras.losses import MeanSquaredError
from keras.optimizers import SGD
from keras.optimizers import Adam
from keras.optimizers import RMSprop, Adadelta, Adagrad, Adamax, Nadam

# optimizer_type = SGD      # this works
optimizer_type = Adam       # this does not work not using checkpoint, works using checkpoint
# optimizer_type = RMSprop      # this does not work not using checkpoint, works using checkpoint
# optimizer_type = Adadelta     # this does not work not using checkpoint, works using checkpoint
# optimizer_type = Adagrad      # this does not work not using checkpoint, works using checkpoint
# optimizer_type = Nadam        # this does not work not using checkpoint, works using checkpoint

np.random.seed(42)

x = np.array(range(100), dtype=np.float32).reshape(-1, 1)
x_batched = np.split(x, 2)
y = 2 * x + 1 + np.random.normal(size=x.shape[0]).reshape(-1, 1)
y_batched = np.split(y, 2)

loss_fn = MeanSquaredError()

class MyModel(Model):
    def __init__(self):
        super().__init__()
        self.d1 = Dense(5, activation='relu', dtype=tf.float32)
        self.d2 = Dense(1, dtype=tf.float32)

    def call(self, inputs):
        return self.d2(self.d1(inputs))


def train_step(model, loss_fn, optimizer, x, y):
    with tf.GradientTape() as tape:
        logits = model(x, training=True)
        loss_value = loss_fn(y, logits)
    grads = tape.gradient(loss_value, model.trainable_weights)
    optimizer.apply_gradients(zip(grads, model.trainable_weights))
    return loss_value


def optimize(model, loss_fn, optimizer, epochs):
    losses = []
    for epoch in range(epochs):
        for x_batch, y_batch in zip(x_batched, y_batched):
            loss_value = train_step(model, loss_fn, optimizer, x_batch,
                                    y_batch)
            losses.append(loss_value.numpy())
    return losses


# train 4 epochs

tf.random.set_seed(42)
model = MyModel()
optimizer = optimizer_type(learning_rate=1e-3)
losses = optimize(model, loss_fn, optimizer, epochs=4)
pred = model(x_batched[0])

# train 2 epochs

tf.random.set_seed(42)
model = MyModel()
optimizer = optimizer_type(learning_rate=1e-3)
checkpoint = tf.train.Checkpoint(model=model, optimizer=optimizer)
manager = tf.train.CheckpointManager(checkpoint, './study/test_save_optimizer/checkpoint', max_to_keep=1)


losses2 = optimize(model, loss_fn, optimizer, epochs=2)

save_path = manager.save()
print("Saved checkpoint {}".format(save_path))

# Previous Trials
# model.save_weights('./study/test_save_optimizer/')
# with open('./study/test_save_optimizer/test_save_optimizer.pkl', 'wb') as f:
#     pickle.dump(optimizer.get_weights(), f)


# load and train 2 more epochs
model = MyModel()
optimizer = optimizer_type(learning_rate=1e-3)
checkpoint = tf.train.Checkpoint(model=model, optimizer=optimizer)
manager = tf.train.CheckpointManager(checkpoint, './study/test_save_optimizer/checkpoint', max_to_keep=1)

# Previous Trials
# model.load_weights('./study/test_save_optimizer/')
# with open('./study/test_save_optimizer/test_save_optimizer.pkl', 'rb') as f:
#     optimizer.set_weights(pickle.load(f))

checkpoint.restore(manager.latest_checkpoint)
if manager.latest_checkpoint:
    print("Restored from {}".format(manager.latest_checkpoint))
else:
    print("Initializing from scratch.")

losses2.extend(optimize(model, loss_fn, optimizer, epochs=2))
pred2 = model(x_batched[0])

print(losses)
print(losses2)
print(pred)
print(pred2)

assert losses == losses2
assert np.all(pred == pred2)
```

**Comment by ooshyun** [comment]
And, I also found that after redefining the optimizer, applying gradient using optimizer at the first time to get the optimizer shape for the optimizer weight's shape, and setting the saved optimizer weight, then the other types of optimizers works well. I attached my test code,

**Comment by k-w-w** [comment]
Using `tf.train.Checkpoint` would be the way of saving the optimizer weights separately from the model (as @ooshyun found)

> Is there any solution or news for getting and setting the optimizer?

@ooshyun  Outside of getting the optimizer with `model.optimizer`, and setting the optimizer with `model.compile`, I'm not sure how to answer this question. Could you provide more details?

**Comment by ooshyun** [comment]
Thanks for answers ! As you can see in the my code, I want to save optimizer state in the custom training model. So it found two ways, first way is using ```tf.train.Checkpoint``` and second way is using get_weights and set_weights. 

The first way worked well, but the second way didn't worked well if you just set the optimizers weights to optimizer.

So, I tried to find the way how to set the weight using get_weights and set_weights. And above sample code is the code I found how to save the optimizer state using getting and setting weights.

However, it needs to do extra works, which were ```model.compile``` and ```model.fit``` before setting the optimizer's weights. So I wanna ask, is there any solution or updates for not doing this extra work? or is there any updates for saving optimizer separately without this two way I mentioned?

**Comment by gowthamkpr** [comment]
@w4nderlust @ooshyun You can take a look at the solutions provided [here](https://stackoverflow.com/questions/49503748/save-and-load-model-optimizer-state/49504376) for saving and loading optimizer state. Thanks!
